src/io/json/read/deserialize.rs

```diff
@@ -243,6 +252,137 @@ fn deserialize_list<'a, O: Offset, A: Borrow<Value<'a>>>(
+    rows: &[A],
+) {
+    let start = {
+        let empty = vec![];
+        let inner: Vec<_> = rows
+            .iter()
+            .flat_map(|row| match row.borrow() {
+                Value::Array(value) => value.iter(),
+                _ => empty.iter(),
+            })
+            .collect();
+
+        let child = target.mut_values();
+        let start_len = child.len();
+        deserialize_into(child, &inner);
+
+        // todo make this an Err
+        O::from_usize(start_len).expect("Child list size too large")
+    };
+
+    let mut position = start;
+    let arrays = rows.iter().map(|row| {
+        match row.borrow() {
+            Value::Array(value) => {
+                // todo make this an Err
+                position += O::from_usize(value.len()).expect("List offset is too large :/");
+                Some(position)
+            }
+            _ => None,
+        }
+    });
+
+    // though this will always be safe, we cannot use unsafe_extend_offsets here
+    // due to `#![forbid(unsafe_code)]` on the io module
+    target.extend_offsets(arrays);
+}
+
+fn deserialize_fixed_size_list_into<'a, A: Borrow<Value<'a>>>(
+    target: &mut MutableFixedSizeListArray<Box<dyn MutableArray>>,
+    rows: &[A],
+) {
+    for row in rows {
+        match row.borrow() {
+            Value::Array(value) => {
+                if value.len() == target.size() {
+                    {
+                        let child = target.mut_values();
+                        deserialize_into(child, value);
+                    }
+                    // unless alignment is already off, the if above should
+                    // prevent this from ever happening.
+                    target.try_push_valid().expect("unaligned backing array");
+                } else {
+                    target.push_null();
+                }
+            }
+            _ => target.push_null(),
+        }
+    }
+}
+
+fn try_deserialize_into<'a, A: Borrow<Value<'a>>, T: NativeType>(
+    target: &mut Box<dyn MutableArray>,
+    rows: &[A],
+    deserialize_into: fn(&mut MutablePrimitiveArray<T>, &[A]) -> (),
+) -> bool {
+    try_generic_deserialize_into(target, rows, deserialize_into)
+}
+
+fn try_generic_deserialize_into<'a, A: Borrow<Value<'a>>, M: 'static>(
+    target: &mut Box<dyn MutableArray>,
+    rows: &[A],
+    deserialize_into: fn(&mut M, &[A]) -> (),
+) -> bool {
+    if let Some(array) = target.as_mut_any().downcast_mut::<M>() {
+        deserialize_into(array, rows);
+        true
+    } else {
+        false
+    }
+}
+
+/// Deserialize `rows` by extending them into the given `target`
+fn deserialize_into<'a, A: Borrow<Value<'a>>>(target: &mut Box<dyn MutableArray>, rows: &[A]) {
+    // It'd be nice to have something like pattern matching for downcasting from Any
+    // I'm not aware of anything like that, which leads to this ... ugliness
+    #[allow(clippy::if_same_then_else)]
+    if let Some(list_array) = target
```

Comment:
we have been using a different approach for these:

```
match self.data_type().to_logical_type() {
    ...
}
```

see e.g. https://github.com/jorgecarleitao/arrow2/blob/main/src/array/equal/mod.rs#L208

could we use this pattern instead?

Reply:
My concern there is we'd still need to do downcasting in each match arm, and then unwrapping. Example:

```
match self.data_type().to_logical_type() {
  // ...
  DataType::Float32 => 
    deserialize_into(target.as_mut_any().downcast_mut::<MutablePrimitiveArray<f32>>().unwrap(), rows),
```

Is there any way to avoid that `.unwrap()` in each arm? It should always succeed, but it feels kinda ugly. Not sure if I'm missing a better way.

Reply:
afaik there is no better way since the invariant "`DataType` is `DataType::Int32` implies `&dyn Array` is `&PrimitiveArray<i32>` and `&dyn MutableArray` is `&MutablePrimitiveArray<i32>`" is a runtime invariant. This enforcement is `safe`, which means that in theory we could even use `.unwrap_unchecked` (but it is not worth the risk imo).